tool/tsh/app_cloud.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/client"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// cloudApp is a dummy interface shared between awsApp, azureApp and gcpApp
+type cloudApp interface {
+	dummyCloudAppMethod()
+}
+
+// cloudAppImpl automatically implements cloudApp when inherited
+type cloudAppImpl struct {
+}
+
+func (_ cloudAppImpl) dummyCloudAppMethod() {
+}
+
+// cloudAppInfo contains common information for given cloudApp
+type cloudAppInfo struct {
+	// matchRouteToApp returns true if given tlsca.RouteToApp is recognized as a given cloud app kind.
+	matchRouteToApp func(tlsca.RouteToApp) bool
+	// newCloudApp creates a new instance of cloudApp
+	newCloudApp func(cf *CLIConf, profile *client.ProfileStatus, appRoute tlsca.RouteToApp) (cloudApp, error)
+	// cloudFriendlyName is a friendly name for given cloud
+	cloudFriendlyName string
+}
+
+func (info *cloudAppInfo) pickActiveCloudApp(cf *CLIConf) (cloudApp, error) {
+	app, needLogin, err := info.pickActiveCloudAppNoRetry(cf)
```

Comment:
What is the context of `NoRetry` can we be more specific  ?

Reply:
What name would you suggest?

Reply:
I woulld just recomend to remove the NoRetry prefix and replace retry term with Relogin like in case of database access https://github.com/gravitational/teleport/blob/4b617352f8912e6c24cc74124d9f4637fa2fccd3/tool/tsh/db.go#L945

Reply:
I have simplified the names.